include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,506 @@
+
+        sycl::group_barrier(__item.get_group());
+    }
+
+    return (__local_id ? __local_acc[__first + __local_id - 1] : 0);
+}
+
+enum class scan_type
+{
+    inclusive,
+    exclusive
+};
+
+template <scan_type __scan_type>
+struct sycl_scan_by_segment_impl
+{
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3,
+              typename _BinaryPredicate, typename _BinaryOperator>
+    void
+    sycl_inclusive_scan_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values,
+                                   _Range3&& __out_values, _BinaryPredicate __binary_pred, _BinaryOperator __binary_op)
+    {
+        using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+        using __key_type = oneapi::dpl::__internal::__value_t<_Range1>;
+        using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+        using __flag_type = bool;
+
+        const int __n = __keys.size();
+
+        constexpr int __vals_per_item = 2; // Assigning 2 elements per work item resulted in best performance on gpu.
+
+        ::std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+
+        // change __wgroup_size according to local memory limit
+        __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(
+            ::std::forward<_ExecutionPolicy>(__exec), sizeof(__key_type) + sizeof(__val_type), __wgroup_size);
+
+        int32_t __n_groups = 1 + std::ceil(__n / (__wgroup_size * __vals_per_item));
+
+        // intermediate scan values stored by each work item
+        auto __partials =
+            oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __val_type>(__exec, __n_groups)
+                .get_buffer();
+
+        auto __end_idx =
+            oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __val_type>(__exec, 1)
+                .get_buffer();
+
+        // the number of segment ends found in each work group
+        auto __seg_ends =
+            oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, bool>(__exec, __n_groups)
+                .get_buffer();
+
+        // 1. Work group scan
+        auto __wg_incl_scan = __exec.queue().submit([&](sycl::handler& __cgh) {
+            auto __partials_acc = __partials.template get_access<sycl::access_mode::write>(__cgh);
+            auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::write>(__cgh);
+
+            oneapi::dpl::__ranges::__require_access(__cgh, __keys, __values, __out_values);
+
+            auto __loc_acc = sycl::accessor<__val_type, 1, sycl::access::mode::read_write, sycl::access::target::local>{
+                2 * __wgroup_size, __cgh};
+
+            __cgh.parallel_for(
+                sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                    __val_type __accumulator = {};
```

Comment:
Additional requirement (default constructor ) on  `__val_type`. Can we avoid that?

Reply:
Investigating this a bit more, this implementation not only requires a default constructor, but also requires it to be the identity for the `__binary_op` and `__val_type`.  We likely should specialize and have two internal implementations, one which takes full advantage of a `__known_identity<__binary_op, __val_type>`, and another which catches the general case without a known identity, the existing scan and reduce algorithms use this strategy.

Reply:
This has been resolved with known_identity special casing for the scan algorithm(s).